### ISSUE TITLE
85% CPU Utiliziation on a Overclocked 10700k

### ISSUE THREAD

**Issue by SentaiBrad** [Skretzo/shortest-path#6]
Talking in the RuneLite discord and it looks like it's well known that the performance of the plugin can eat a CPU. Even trying to adjust the settings so it doesn't re-calculate often it's pinging my CPU and crashing my game when making paths longer than 20 cells, especially around Varrock.

Possibility for this to be more efficient? Any information I can provide to narrow down what could be causing this?

**Comment by Skretzo** [comment]
Does this only happen when your path ends on a tile which you cannot reach (e.g. inside a tree or unreachable area)?

**Comment by SentaiBrad** [comment]
Randomly. It doesn't matter where I palce it.

**Comment by Skretzo** [comment]
Hmm, this is very weird as your provided image and path has a target tile that is not unreachable (which would have been an issue).

Is the path red or blue when your issue occurs?
(red is the default colour for when the path is finished, while blue is the default colour for when the path is searching/calculating and is not yet finished).

Would it be possible for you to record a video of it happening? If so, please have the FPS plugin enabled so that I can see when your PC is struggling.
(you should be able to drag & drop a MP4 video file on the comment text field to send it to me, alternatively you can send it to me on discord Skretzo#1303)

**Comment by SentaiBrad** [comment]
It has happened when the line is both colors, but the video shows blue. It does happen a lot more around complicated Terrain or bigger cities. 

GitHub wanted under 10mb for a video xD https://youtu.be/Z3ZrGUdL97E

**Comment by SentaiBrad** [comment]
No worries at all. The fast you responded fast and are open to figuring it out is awesome enough. Some devs don't give a shit xD. If I could have helped even a little, I am happy.

**Comment by vbojilov** [comment]
I just wanted to say I have this issue as well, and it makes turning the plugin on, a death sentence. One thing I wanted to potentially add is that while I don't have proof or direct comparison, I don't remember having issues anywhere this severe prior to the recent slew of Runelite updates. In the past, I've used this plugin to route cross-map, from the GE to Port Sarim as a specific example, without an issue, but now any distance causes performance drops with even paths contained in a single chunk sometimes necessitating a force close due to bogged performance and freeze.

**Comment by SentaiBrad** [comment]
> I just wanted to say I have this issue as well, and it makes turning the plugin on, a death sentence. One thing I wanted to potentially add is that while I don't have proof, I don't remember having issues anywhere this severe prior to the recent slew of Runelite updates. In the past, I've used this plugin to route cross-map, from the GE to Port Sarim as a specific example, without an issue, but now any distance causes performance drops with even paths contained in a single chunk sometimes necessitating a force close due to bogged performance and freeze.

To that end, I don't recall a performance loss when I last used this heavily a year ago.

**Comment by Skretzo** [comment]
> I just wanted to say I have this issue as well, and it makes turning the plugin on, a death sentence.

@vbojilov Could you elaborate on how turning the plugin on is affecting your performance? Is that after having previously had the plugin turned on and made a path that caused a performance drop?

Also, can either of you check if the `.log` files in your `%userprofile%\.runelite\logs` contain any information of a crash/error related to the `shortest-path` plugin? You might have to reproduce the error to update the log file.

**Comment by vbojilov** [comment]
I'm sorry, to clarify- I don't believe I'm seeing apparent performance problems with the plugin simply toggled on, but when a path is selected the performance cascades down after 3-5 seconds or so. 

I just reproduced the issue at least twice with a blue marked path. (One attempt in between generated a red path which seemed to work fine, however I know that I've had problems with red paths recently so I'll continue testing to see if I can reproduce that as well.) The issue progresses whether I follow the path (game loads new areas) or whether I don't move at all. 

Here's what it looks like: I'm on a mobile APU with 16 GB of RAM. It's not the most powerful of systems but it's plenty to run the game relatively consistently at 45-50 FPS depending on area. When a path is marked, the FPS indicator shows performance mostly in line with the usual, 40-45ish frames per second. After a few seconds, though, the numbers start going down, first to the 30s, then after another second or two to the 20s, then the game effectively locks itself as the FPS stays stuck on a low single digit (1 or 2) with the screen frozen. The PC itself doesn't freeze, mouse movement works normally, other programs can be operated without a problem, etc. Runelite stops responding. If I right click the minimap to clear path while the game still processes inputs (10s to 20s FPS) then depending on how late I waited to clear the path it will still freeze but then it recovers. Throughout the fall of FPS, the game acts very stuttery with microfreezes/lag throughout the entire time the path is marked. So 20 or 30 FPS doesn't feel like 20 or 30, it feels much worse. 

Useful to you might be the portion of the log file where I found errors during the time I was testing! If you need the full error log for whatever reason, let me know. However this encapsulates the full period where I was logged in and testing this issue.
<details>
  <summary>Error Log portion</summary>

  ```
2022-07-10 12:11:35 [pool-8-thread-1] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.addConditionWaiter(Unknown Source)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.getTask(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2022-07-10 12:11:55 [Thread-55] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
2022-07-10 12:11:55 [pool-6-thread-1] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
2022-07-10 12:11:55 [pool-3-thread-1] WARN  n.runelite.client.game.WorldService - Error looking up worlds
java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:678)
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:686)
	at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.java:154)
	at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.java:136)
	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at net.runelite.client.RuneLite.lambda$buildHttpClient$5(RuneLite.java:450)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at net.runelite.client.RuneLite.lambda$buildHttpClient$4(RuneLite.java:443)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229)
	at okhttp3.RealCall.execute(RealCall.java:81)
	at net.runelite.client.game.WorldClient.lookupWorlds(WorldClient.java:61)
	at net.runelite.client.game.WorldService.fetch(WorldService.java:96)
	at net.runelite.client.game.WorldService.tick(WorldService.java:81)
	at net.runelite.client.util.RunnableExceptionLogger.run(RunnableExceptionLogger.java:41)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
2022-07-10 12:11:55 [OkHttp ConnectionPool] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
2022-07-10 12:11:55 [D3D Screen Updater] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
2022-07-10 12:11:55 [pool-11-thread-1] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space
2022-07-10 12:11:55 [Client] ERROR net.runelite.client.RuneLite - Uncaught exception:
java.lang.OutOfMemoryError: Java heap space: failed reallocation of scalar replaced objects
```
</details>


Also, thank you for your development. It's very much appreciated!

**Comment by idle-user** [comment]
Also getting the same lagging/freezing issue randomly when setting a path. Client hangs terribly and stops responding. Have to kill the task usually. I checked runelite logs and I cannot grep relevant logs matching "shortest-path". 

32GB RAM and R7 3800x CPU

If there's any way I can help by searching other keywords, I'd be glad to help.
Thank you,

**Comment by Skretzo** [comment]
@idle-user I don't think there is any need to look for logs anymore. I just need to rewrite the pathfinding algorithm itself, which I am currently working on now.

**Comment by Skretzo** [comment]
I have released a preliminary fix that is now live in the client, and would be happy if I could get feedback om whether the issue is solved or not. Just restart your client to fetch the update if you have not done so in the last 6 hours or so.
What I have done is to stop the pathfinding if no new progress has been made for 2 seconds.

**Comment by SentaiBrad** [comment]
I have tried several complicated paths, some behind locked doors or paths that require going into caves, etc. While drawing and pathing it still lags, but once it's placed it is fine. Short paths seem to be good as well. So the only lag (so far) seems to be when long paths are being drawn, but that is probably expected.

**Comment by Skretzo** [comment]
Thanks to major performance improvements by @jocopa3, this plugin is now much faster and uses less memory. These improvements were released in version 1.10, which is now live in the client. If there are still performance issues, please let me know by either re-opening this issue or creating a new one.
